jerry-core/parser/regexp/re-compiler.c

```diff
@@ -533,90 +545,93 @@ re_compile_bytecode (const re_compiled_code_t **out_bytecode_p, /**< [out] point
+    re_append_opcode (&bc_ctx, RE_OP_EOF);
+
+    /* 3. Insert extra informations for bytecode header */
+    re_compiled_code_t re_compiled_code;
+
+    re_compiled_code.flags = re_ctx.flags | (1u << ECMA_BYTECODE_REF_SHIFT);
+    ECMA_SET_NON_NULL_POINTER (re_compiled_code.pattern_cp,
+                               ecma_copy_or_ref_ecma_string (pattern_str_p));
+    re_compiled_code.num_of_captures = re_ctx.num_of_captures * 2;
+    re_compiled_code.num_of_non_captures = re_ctx.num_of_non_captures;
+
+    re_bytecode_list_insert (&bc_ctx,
+                             0,
+                             (uint8_t *) &re_compiled_code,
+                             sizeof (re_compiled_code_t));
+  }
 
-    /* 2. Check for invalid backreference */
-    if (re_ctx.highest_backref >= re_ctx.num_of_captures)
-    {
-      ret_value = ecma_raise_syntax_error ("Invalid backreference.\n");
-    }
-    else
-    {
-      re_append_opcode (&bc_ctx, RE_OP_SAVE_AND_MATCH);
-      re_append_opcode (&bc_ctx, RE_OP_EOF);
-
-      /* 3. Insert extra informations for bytecode header */
-      re_compiled_code_t re_compiled_code;
-
-      re_compiled_code.flags = re_ctx.flags | (1u << ECMA_BYTECODE_REF_SHIFT);
-      ECMA_SET_NON_NULL_POINTER (re_compiled_code.pattern_cp,
-                                 ecma_copy_or_ref_ecma_string (pattern_str_p));
-      re_compiled_code.num_of_captures = re_ctx.num_of_captures * 2;
-      re_compiled_code.num_of_non_captures = re_ctx.num_of_non_captures;
-
-      re_bytecode_list_insert (&bc_ctx,
-                               0,
-                               (uint8_t *) &re_compiled_code,
-                               sizeof (re_compiled_code_t));
-    }
+  ECMA_FINALIZE (empty);
 
-    ECMA_FINALIZE (empty);
+  MEM_FINALIZE_LOCAL_ARRAY (pattern_start_p);
 
-    MEM_FINALIZE_LOCAL_ARRAY (pattern_start_p);
+  if (!ecma_is_value_empty (ret_value))
+  {
+    /* Compilation failed, free bytecode. */
+    JERRY_DDLOG ("RegExp compilation failed!\n");
+    mem_heap_free_block_size_stored (bc_ctx.block_start_p);
+    *out_bytecode_p = NULL;
+  }
+  else
+  {
+#ifdef JERRY_ENABLE_LOG
+    re_dump_bytecode (&bc_ctx);
+#endif
 
-    if (!ecma_is_value_empty (ret_value))
-    {
-      /* Compilation failed, free bytecode. */
-      mem_heap_free_block_size_stored (bc_ctx.block_start_p);
-      *out_bytecode_p = NULL;
-    }
-    else
-    {
-      /* The RegExp bytecode contains at least a RE_OP_SAVE_AT_START opdoce, so it cannot be NULL. */
-      JERRY_ASSERT (bc_ctx.block_start_p != NULL);
-      *out_bytecode_p = (re_compiled_code_t *) bc_ctx.block_start_p;
+    /* The RegExp bytecode contains at least a RE_OP_SAVE_AT_START opdoce, so it cannot be NULL. */
+    JERRY_ASSERT (bc_ctx.block_start_p != NULL);
+    *out_bytecode_p = (re_compiled_code_t *) bc_ctx.block_start_p;
 
-      if (cache_idx < RE_CACHE_SIZE)
+    if (cache_idx >= RE_CACHE_SIZE)
+    {
+      if (re_cache_idx < 1u)
```

Comment:
Actually, in these two `if`s, we are checking whether `cache_idx == RE_CACHE_SIZE` and `re_cache_idx == 0`, aren't we? (I mean, there is no way for `cache_idx` to be striclty larger that `RE_CACHE_SIZE`, and `re_cache_idx` being unsigned, it cannot have negative value.) I'd vote for the exact equality tests.